sigstore-java/src/main/java/dev/sigstore/tuf/Updater.java

```diff
@@ -225,82 +234,190 @@ void verifyDelegate(
     }
   }
 
-  Timestamp updateTimestamp(Root root)
+  Optional<Timestamp> updateTimestamp(Root root)
       throws IOException, NoSuchAlgorithmException, InvalidKeySpecException, InvalidKeyException,
-          MetaNotFoundException, SignatureVerificationException {
+          FileNotFoundException, SignatureVerificationException {
     // 1) download the timestamp.json bytes.
     var timestamp =
         fetcher
             .getMeta(Role.Name.TIMESTAMP, Timestamp.class)
-            .orElseThrow(
-                () -> new MetaNotFoundException("could not find timestamp.json on mirror."))
+            .orElseThrow(() -> new FileNotFoundException("timestamp.json", fetcher.getSource()))
             .getMetaResource();
 
     // 2) verify against threshold of keys as specified in trusted root.json
     verifyDelegate(root, timestamp);
 
-    // 3) check that version of new timestamp.json is higher or equal than current, else fail.
-    //     3.2) check that timestamp.snapshot.version <= timestamp.version or fail
+    // 3) If the new timestamp file has a lesser version than our current trusted timestamp file
+    // report a rollback attack.  If it is equal abort the update as there should be no changes. If
+    // it is higher than continue update.
     Optional<Timestamp> localTimestampMaybe = localStore.loadTimestamp();
     if (localTimestampMaybe.isPresent()) {
       Timestamp localTimestamp = localTimestampMaybe.get();
       if (localTimestampMaybe.get().getSignedMeta().getVersion()
-          >= timestamp.getSignedMeta().getVersion()) {
+          > timestamp.getSignedMeta().getVersion()) {
         throw new RollbackVersionException(
             localTimestamp.getSignedMeta().getVersion(), timestamp.getSignedMeta().getVersion());
       }
+      if (localTimestampMaybe.get().getSignedMeta().getVersion()
+          == timestamp.getSignedMeta().getVersion()) {
+        return Optional.empty();
+      }
     }
     // 4) check expiration timestamp is after tuf update start time, else fail.
     throwIfExpired(timestamp.getSignedMeta().getExpiresAsDate());
     // 5) persist timestamp.json
     localStore.storeMeta(timestamp);
-    return timestamp;
+    return Optional.of(timestamp);
   }
 
-  Snapshot updateSnapshot(Root root, Timestamp timestamp) {
+  Snapshot updateSnapshot(Root root, Timestamp timestamp)
+      throws IOException, FileNotFoundException, InvalidHashesException,
+          SignatureVerificationException, NoSuchAlgorithmException, InvalidKeySpecException,
+          InvalidKeyException {
     // 1) download the snapshot.json bytes up to few 10s of K max.
-
+    // TODO:asraa@ to review whether we need to be using the snapshot size from the timestamp file
```

Comment:
@asraa

Reply:
Yes! You can use the number of bytes int the timestamp file.

Reply:
updated to use max bytes to limit snapshot and targets download.